api-participant/src/main/java/bio/terra/pearl/api/participant/service/ExternalEventExtService.java

```diff
@@ -0,0 +1,94 @@
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+
+@Service
+@Slf4j
+public class ExternalEventExtService {
+  private final AuthUtilService authUtilService;
+  private final EnrollmentService enrollmentService;
+  private final PortalEnvironmentService portalEnvironmentService;
+  private final RegistrationService registrationService;
+  private final StudyEnvironmentService studyEnvironmentService;
+  private final EnrolleeService enrolleeService;
+  private final ParticipantUserService participantUserService;
+  private final PortalParticipantUserService portalParticipantUserService;
+  private final EventService eventService;
+
+  public ExternalEventExtService(
+      AuthUtilService authUtilService,
+      EnrollmentService enrollmentService,
+      PortalEnvironmentService portalEnvironmentService,
+      RegistrationService registrationService,
+      StudyEnvironmentService studyEnvironmentService,
+      EnrolleeService enrolleeService,
+      ParticipantUserService participantUserService,
+      PortalParticipantUserService portalParticipantUserService,
+      EventService eventService) {
+    this.authUtilService = authUtilService;
+    this.enrollmentService = enrollmentService;
+    this.portalEnvironmentService = portalEnvironmentService;
+    this.registrationService = registrationService;
+    this.studyEnvironmentService = studyEnvironmentService;
+    this.enrolleeService = enrolleeService;
+    this.participantUserService = participantUserService;
+    this.portalParticipantUserService = portalParticipantUserService;
+    this.eventService = eventService;
+  }
+
+  /**
+   * Track a failed login event asynchronously. Since events must be attached to an enrollee, only
+   * tracks failed logins for registered users.
+   */
+  @Async
+  public void trackFailedLoginEventAsync(
+      String portalShortcode, EnvironmentName envName, String username) {
+    try {
+      trackFailedLogin(portalShortcode, envName, username);
+    } catch (Exception e) {
+      log.error(
+          "Error tracking failed login for user {} in portal {} on environment {}: {}",
+          username,
+          portalShortcode,
+          envName,
+          e.getMessage(),
+          e);
+    }
+  }
+
+  public void trackFailedLogin(String portalShortcode, EnvironmentName envName, String username) {
+
+    Optional<PortalParticipantUser> ppUserOpt =
+        portalParticipantUserService.findOne(username, portalShortcode, envName);
+
+    if (ppUserOpt.isEmpty()) {
+      return; // user not in portal, nothing to track
+    }
+    PortalParticipantUser ppUser = ppUserOpt.get();
+
+    List<Enrollee> enrollees = enrolleeService.findByPortalParticipantUser(ppUser);
+
+    if (enrollees.isEmpty()) {
+      return; // no enrollee to track
+    }
+
+    for (Enrollee enrollee : enrollees) {
+      eventService.publishEnrolleeFailedLoginEvent(enrollee, ppUser);
+    }
```

Comment:
should work just fine for proxies - the failed logins will just be attached to the proxy enrollee instead of governed. which, to me, seems correct.